signac/synced_collections/buffers/buffered_collection.py

```diff
@@ -0,0 +1,187 @@
+    thread safe for the resulting collection type to be thread safe.
+
+    """
+
+    def __init__(self, *args, **kwargs):
+        # The `_buffered` attribute _must_ be defined prior to calling the
+        # superclass constructors in order to enable subclasses to override
+        # setattr and getattr in nontrivial ways. In particular, if setattr and
+        # getattr need to access the synced data, they may call sync and load,
+        # which depend on this parameter existing and could otherwise end up in
+        # an infinite recursion.
+        self.buffered = _CounterFuncContext(self._flush)
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def __init_subclass__(cls):
+        """Register subclasses for the purpose of global buffering.
+
+        Each subclass has its own means of buffering and must be flushed.
+        """
+        super().__init_subclass__()
+        if not isabstract(cls):
+            cls._buffer_context = _CounterFuncContext(cls._flush_buffer)
+
+    @classmethod
+    def buffer_backend(cls, *args, **kwargs):
+        """Enter context to buffer all operations for this backend."""
+        return cls._buffer_context
+
+    @classmethod
+    def backend_is_buffered(cls):
+        """Check if this backend is currently buffered."""
+        return bool(cls._buffer_context)
+
+    def _save(self):
+        """Synchronize data with the backend but buffer if needed.
+
+        This method is identical to the SyncedCollection implementation for
+        `sync` except that it determines whether data is actually synchronized
+        or instead written to a temporary buffer based on the buffering mode.
+        """
+        if not self._suspend_sync:
+            if self._root is None:
+                if self._is_buffered:
+                    self._save_to_buffer()
+                else:
+                    self._save_to_resource()
+            else:
+                self._root._save()
+
+    def _load(self):
+        """Load data from the backend but buffer if needed.
+
+        This method is identical to the :class:`~.SyncedCollection`
+        implementation except that it determines whether data is actually
+        synchronized or instead read from a temporary buffer based on the
+        buffering mode.
+        """
+        if not self._suspend_sync:
+            if self._root is None:
+                if self._is_buffered:
+                    data = self._load_from_buffer()
+                else:
+                    data = self._load_from_resource()
+                with self._suspend_sync:
+                    self._update(data)
+            else:
+                self._root._load()
+
+    def _save_to_buffer(self):
+        """Store data in buffer.
+
+        By default, this method simply calls :meth:`~._save_to_resource`. Subclasses
+        must implement specific buffering strategies.
+        """
+        self._save_to_resource()
```

Comment:
You're right, that's what's missing, but I don't really want to write a new class just for testing something that we don't need right now. We can come back to it whenever we get around to enabling the interchange of backends within signac, since that will require buffered versions of whatever collection type that is.